File: src/main/java/phoenix/init/PhoenixItems.java
package phoenix.init;

import net.minecraft.item.AxeItem;
import net.minecraft.item.Item;
import net.minecraft.item.PickaxeItem;
import net.minecraft.item.SwordItem;
import net.minecraftforge.fml.RegistryObject;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;
import phoenix.Phoenix;
import phoenix.items.ItemDiary;
import phoenix.items.ash.CrucibleItem;
import phoenix.items.ash.FormItem;
import phoenix.items.ash.HighQualityClayItem;
import phoenix.items.ash.KnifeItem;

import java.util.function.Supplier;

public class PhoenixItems
{
    private static final DeferredRegister<Item> ITEMS = DeferredRegister.create(ForgeRegistries.ITEMS, Phoenix.MOD_ID);

    public static final RegistryObject<Item> GUIDE                   = ITEMS.register("diary",            ItemDiary::new);
    public static final RegistryObject<Item> HIGH_QUALITY_CLAY_ITEM  = ITEMS.register("high_quality_clay",HighQualityClayItem::new);
    public static final RegistryObject<Item> CRUCIBLE                = ITEMS.register("crucible",         CrucibleItem::new);

    public static final RegistryObject<Item> CRUCIBLE_WITH_IRON_ORE  = ITEMS.register("crucible_with_iron_ore", CrucibleItem::new);
    public static final RegistryObject<Item> CRUCIBLE_WITH_IRON      = ITEMS.register("crucible_with_iron",     CrucibleItem::new);
    public static final RegistryObject<Item> CRUCIBLE_WITH_GOLD_ORE  = ITEMS.register("crucible_with_gold_ore", CrucibleItem::new);
    public static final RegistryObject<Item> CRUCIBLE_WITH_GOLD      = ITEMS.register("crucible_with_gold",     CrucibleItem::new);

    public static final RegistryObject<Item> STEEL_INGOT             = ITEMS.register("steel_ingot",     basicItem());
    public static final RegistryObject<Item> ZIRCONIUM_INGOT         = ITEMS.register("zirconium_ingot", basicItem());

    public static final RegistryObject<Item> ZIRCONIUM_SWORD_BLADE   = ITEMS.register("zirconium_sword_blade",   basicItem());
    public static final RegistryObject<Item> ZIRCONIUM_KNIFE_BLADE   = ITEMS.register("zirconium_knife_blade",   basicItem());
    public static final RegistryObject<Item> ZIRCONIUM_BAYONET       = ITEMS.register("zirconium_bayonet",       basicItem());
    public static final RegistryObject<Item> ZIRCONIUM_BUTT          = ITEMS.register("zirconium_butt",          basicItem());


    public static final RegistryObject<Item> STEEL_FORM_KNIFE_BLADE  = ITEMS.register("steel_form_knife_blade",   basicItem());
    public static final RegistryObject<Item> STEEL_FORM_SWORD_BLADE  = ITEMS.register("steel_form_sword_blade",   basicItem());
    public static final RegistryObject<Item> STEEl_FORM_BAYONET      = ITEMS.register("steel_form_bayonet",       basicItem());
    public static final RegistryObject<Item> STEEl_FORM_BUTT         = ITEMS.register("steel_form_butt",          basicItem());

    public static final RegistryObject<Item> STEEL_FORM_KNIFE_BLADE_FULL    = ITEMS.register("steel_form_knife_blade_full",   form(STEEL_FORM_KNIFE_BLADE));
    public static final RegistryObject<Item> STEEL_FORM_SWORD_BLADE_FULL    = ITEMS.register("steel_form_sword_blade_full",   form(STEEL_FORM_SWORD_BLADE));
    public static final RegistryObject<Item> STEEl_FORM_BAYONET_FULL        = ITEMS.register("steel_form_bayonet_full",       form(STEEl_FORM_BAYONET));
    public static final RegistryObject<Item> STEEl_FORM_BUTT_FULL           = ITEMS.register("steel_form_butt_full",          form(STEEl_FORM_BUTT));

    public static final RegistryObject<Item> STEEL_FORM_KNIFE_BLADE_ROASTED = ITEMS.register("steel_form_knife_blade_roasted",   form(STEEL_FORM_KNIFE_BLADE));
    public static final RegistryObject<Item> STEEL_FORM_SWORD_BLADE_ROASTED = ITEMS.register("steel_form_sword_blade_roasted",   form(STEEL_FORM_SWORD_BLADE));
    public static final RegistryObject<Item> STEEl_FORM_BAYONET_ROASTED     = ITEMS.register("steel_form_bayonet_roasted",       form(STEEl_FORM_BAYONET));
    public static final RegistryObject<Item> STEEl_FORM_BUTT_ROASTED        = ITEMS.register("steel_form_butt_roasted",          form(STEEl_FORM_BUTT));


    public static final RegistryObject<Item> ZIRCONIUM_AXE          = ITEMS.register("ceramic_zirconium_axe",      () -> new AxeItem(PhoenixTiers.ZIRCONIUM_TIER, 9.0F, 1.1F, new Item.Properties().group(Phoenix.PHOENIX)));
    public static final RegistryObject<Item> ZIRCONIUM_PICKAXE      = ITEMS.register("ceramic_zirconium_pickaxe",  () -> new PickaxeItem(PhoenixTiers.ZIRCONIUM_TIER, 4, 0.5F, new Item.Properties().group(Phoenix.PHOENIX)));
    public static final RegistryObject<Item> ZIRCONIUM_SWORD        = ITEMS.register("ceramic_zirconium_sword",    () -> new SwordItem(PhoenixTiers.ZIRCONIUM_TIER, 3, 1F, new Item.Properties().group(Phoenix.PHOENIX)));
    public static final RegistryObject<KnifeItem> ZIRCONIUM_KNIFE   = ITEMS.register("ceramic_zirconium_knife",    () -> new KnifeItem(PhoenixTiers.ZIRCONIUM_TIER, 3, 1F, PhoenixConfiguration.COMMON_CONFIG.HARDCORE.get().maxKnifeUsages));

    public static void register()
    {
        ITEMS.register(FMLJavaModLoadingContext.get().getModEventBus());
    }

    private static Supplier<? extends Item> basicItem()
    {
        return () -> new Item(new Item.Properties().group(Phoenix.PHOENIX));
    }

    private static Supplier<? extends Item> form(RegistryObject<Item> contains)
    {
        return () -> new Item(new Item.Properties().group(Phoenix.PHOENIX).containerItem(contains.get()));
    }

    private static Supplier<? extends Item> form(Item contains)
    {
        return () -> new Item(new Item.Properties().group(Phoenix.PHOENIX).containerItem(contains));
    }
}

File: src/main/kotlin/phoenix/init/PhoenixItems.kt
package phoenix.init

import net.minecraft.entity.Entity
import net.minecraft.inventory.EquipmentSlotType
import net.minecraft.item.*
import net.minecraftforge.registries.ForgeRegistries
import phoenix.Phoenix
import phoenix.Phoenix.Companion.ASH
import phoenix.items.ItemDiary
import phoenix.items.ash.CrucibleItem
import phoenix.items.ash.KnifeItem
import thedarkcolour.kotlinforforge.forge.KDeferredRegister
import thedarkcolour.kotlinforforge.forge.MOD_BUS

object PhoenixItems
{
    private val ITEMS = KDeferredRegister(ForgeRegistries.ITEMS, Phoenix.MOD_ID)

    val GUIDE                          by ITEMS.register("diary", ::ItemDiary)

    val HIGH_QUALITY_CLAY_ITEM         by ITEMS.register("high_quality_clay", basicItem(Item.Properties().maxStackSize(16).group(ASH)))
    val COOKED_SETA                    by ITEMS.register("cooked_seta", basicFood(Food.Builder().hunger(3).saturation(1.0f).fastToEat().build(), ItemGroup.FOOD))!!
    val GOLDEN_SETA                    by ITEMS.register("golden_seta", basicFood(Food.Builder().hunger(5).saturation(1.5F).fastToEat().build(), ItemGroup.FOOD))!!

    val CRUCIBLE                       by ITEMS.register("crucible", basicItem())
    val CRUCIBLE_WITH_IRON_ORE         by ITEMS.register("crucible_with_iron_ore", ::CrucibleItem)
    val CRUCIBLE_WITH_IRON             by ITEMS.register("crucible_with_steel",     ::CrucibleItem)
    val CRUCIBLE_WITH_GOLD_ORE         by ITEMS.register("crucible_with_gold_ore", ::CrucibleItem)
    val CRUCIBLE_WITH_GOLD             by ITEMS.register("crucible_with_gold",     ::CrucibleItem)

    val IRON_ORE_WITH_COAL             by ITEMS.register("iron_ore_with_coal", basicItem())
    val IRON_WITH_COAL                 by ITEMS.register("iron_with_coal",     basicItem())
    val STEEL_INGOT                    by ITEMS.register("steel_ingot",        basicItem())
    val ZIRCONIUM_INGOT                by ITEMS.register("zirconium_ingot",    basicItem())

    val ZIRCONIUM_SWORD_BLADE          by ITEMS.register("zirconium_sword_blade", basicItem())
    val ZIRCONIUM_KNIFE_BLADE          by ITEMS.register("zirconium_knife_blade", basicItem())
    val ZIRCONIUM_BAYONET              by ITEMS.register("zirconium_pickaxe_head",basicItem())
    val ZIRCONIUM_AXE_BLADE            by ITEMS.register("zirconium_axe_blade",   basicItem())

    val STEEL_FORM_KNIFE_BLADE         by ITEMS.register("steel_form_knife_blade", basicItem())
    val STEEL_FORM_SWORD_BLADE         by ITEMS.register("steel_form_sword_blade", basicItem())
    val STEEl_FORM_PICKAXE_HEAD        by ITEMS.register("steel_form_pickaxe_head",basicItem())
    val STEEl_FORM_AXE_BLADE           by ITEMS.register("steel_form_axe_blade",   basicItem())

    val STEEL_FORM_KNIFE_BLADE_FULL    by ITEMS.register("steel_form_knife_blade_full", form { STEEL_FORM_KNIFE_BLADE })
    val STEEL_FORM_SWORD_BLADE_FULL    by ITEMS.register("steel_form_sword_blade_full", form { STEEL_FORM_SWORD_BLADE })
    val STEEl_FORM_PICKAXE_HEAD_FULL   by ITEMS.register("steel_form_pickaxe_head_full",form { STEEl_FORM_PICKAXE_HEAD })
    val STEEl_FORM_AXE_BLADE_FULL      by ITEMS.register("steel_form_axe_blade_full",   form { STEEl_FORM_AXE_BLADE })

    val STEEL_FORM_KNIFE_BLADE_ROASTED  by ITEMS.register("steel_form_knife_blade_roasted", form {STEEL_FORM_KNIFE_BLADE})
    val STEEL_FORM_SWORD_BLADE_ROASTED  by ITEMS.register("steel_form_sword_blade_roasted", form {STEEL_FORM_SWORD_BLADE})
    val STEEl_FORM_PICKAXE_HEAD_ROASTED by ITEMS.register("steel_form_pickaxe_head_roasted",form {STEEl_FORM_PICKAXE_HEAD})
    val STEEl_FORM_AXE_BLADE_ROASTED    by ITEMS.register("steel_form_axe_blade_roasted",   form {STEEl_FORM_AXE_BLADE})

    val ZIRCONIUM_AXE                  by ITEMS.register("ceramic_zirconium_axe")     { AxeItem    (PhoenixTiers.ZIRCONIUM_TIER, 9.0f, -3.2f, Item.Properties().group(ASH)) }
    val ZIRCONIUM_PICKAXE              by ITEMS.register("ceramic_zirconium_pickaxe") { PickaxeItem(PhoenixTiers.ZIRCONIUM_TIER, 0, -1f, Item.Properties().group(ASH)) }
    val ZIRCONIUM_SWORD                by ITEMS.register("ceramic_zirconium_sword")   { SwordItem  (PhoenixTiers.ZIRCONIUM_TIER, 2, -0.5f, Item.Properties().group(ASH)) }
    val ZIRCONIUM_KNIFE                by ITEMS.register("ceramic_zirconium_knife")   { KnifeItem  (PhoenixTiers.ZIRCONIUM_TIER, 3f, -10f, PhoenixConfiguration.COMMON_CONFIG.gameMode.get().maxKnifeUsages, ASH) }

    val STEEL_AXE                      by ITEMS.register("steel_axe")     { AxeItem    (PhoenixTiers.STEEL_TIER, 5.0f, -2f, Item.Properties().group(ASH)) }
    val STEEL_PICKAXE                  by ITEMS.register("steel_pickaxe") { PickaxeItem(PhoenixTiers.STEEL_TIER, 0,    -2f, Item.Properties().group(ASH)) }
    val STEEL_SWORD                    by ITEMS.register("steel_sword")   { SwordItem  (PhoenixTiers.STEEL_TIER, 2,    -2f, Item.Properties().group(ASH)) }

    val STEEL_ARMOR_HEAD               by ITEMS.register("steel_armor_head")      { SteelArmorItem(EquipmentSlotType.HEAD, Item.Properties().group(ASH)) }
    val STEEL_ARMOR_CHES               by ITEMS.register("steel_armor_chestplate"){ SteelArmorItem(EquipmentSlotType.CHEST,Item.Properties().group(ASH)) }
    val STEEL_ARMOR_LEGG               by ITEMS.register("steel_armor_leggings")  { SteelArmorItem(EquipmentSlotType.LEGS, Item.Properties().group(ASH)) }
    val STEEL_ARMOR_BUTS               by ITEMS.register("steel_armor_boots")     { SteelArmorItem(EquipmentSlotType.FEET, Item.Properties().group(ASH)) }

    fun register() = ITEMS.register(MOD_BUS)

    private fun basicItem()                             : () -> Item  = { Item(Item.Properties().group(ASH))                               }
    private fun basicItem(prop : Item.Properties)       : () -> Item  = { Item(prop)                                                       }
    private fun basicFood(food: Food)                   : () -> Item  = { Item(Item.Properties().group(ASH).food(food))                    }
    private fun basicFood(food: Food, group : ItemGroup): () -> Item  = { Item(Item.Properties().group(group).food(food))                  }
    private fun form(contains: () -> Item)                    : () -> Item  = { Item(Item.Properties().group(ASH).containerItem(contains())) }
}

class SteelArmorItem(slot: EquipmentSlotType, builder: Properties) : ArmorItem(PhoenixArmorMaterials.STEEL, slot, builder)
{
    override fun getArmorTexture(
        itemstack: ItemStack?,
        entity: Entity?,
        slot: EquipmentSlotType?,
        layer: String?
    ): String
    {
        return if (slot == EquipmentSlotType.LEGS)
        {
            Phoenix.MOD_ID + ":textures/models/armor/steel_layer_2.png"
        } else
        {
            Phoenix.MOD_ID + ":textures/models/armor/steel_layer_1.png"
        }
    }
}